teps/0048-task-results-without-results.md

```diff
@@ -0,0 +1,293 @@
+---
+title: Task Results without Results
+authors:
+  - "@pritidesai"
+  - "@jerop"
+creation-date: 2020-10-20
+last-updated: 2021-06-11
+status: proposed
+---
+
+# TEP-0048: Task Results without Results
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Requirements](#requirements)
+  - [Use Cases](#use-cases)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+A `task` in a `pipeline` can produce a result and that result can be consumed in many ways within that `pipeline`:
+
+* `params` mapping in a consumer `pipelineTask`
+
+```yaml
+kind: Pipeline
+spec:
+  tasks:
+    - name: format-result
+      taskRef:
+        name: format-result
+      params:
+        - name: result
+          value: "$(tasks.sum-inputs.results.result)"
+```
+
+* `WhenExpressions`
+
+```yaml
+kind: Pipeline
+spec:
+  tasks:
+    - name: echo-file-exists
+      when:
+        - input: "$(tasks.check-file.results.exists)"
+          operator: in
+          values: ["yes"]
+```
+
+* Pipeline Results
+
+```yaml
+kind: Pipeline
+spec:
+  tasks:
+    ...
+  results:
+    - name: sum
+      value: $(tasks.second-add.results.sum)
+```
+
+Today, `pipeline` is declared `failure` and stops executing further if the result resolution fails because of
+a missing task result. There are many reasons for a missing task result:
+ 
+* a `task` producing task result failed, no result available
+* a `task` producing result was skipped/disabled and no result generated
+* a `task` producing result did not generate that result even without any failure. We have a
+  [bug report](https://github.com/tektoncd/pipeline/issues/3497) open to declare
+  such a task as failure. This reason might not hold true after issue
+  [#3497]((https://github.com/tektoncd/pipeline/issues/3497)) is fixed.
+
+Here are the major motivations for `pipeline` authors to design their pipelines with the missing task results:
+
+* Implementing the [TEP-0059: Skipping Strategies](0059-skipping-strategies.md)
+  proposal to limit the scope of `WhenExpressions` to only that task and continue executing the dependencies.
+
+  Let's revisit an [example](0059-skipping-strategies.md#use-cases) of sending a Slack notification when someone
+  manually approves the PR. This is done by sending the `approver`'s name to the `slack-msg` task as the result of
+  `manual-approval` task.
+
+  Further, extending the same use case, when someone approves the PR, the `approver` would be set to an appropriate
+  name. At the same time, set the task result `approver` to **None** in case the `manual-approval` task is skipped and
+  the `approver` is not initialized. It is still possible to send a notification that no one approved the PR.
```

Comment:
hmm interesting - in TEP-0059 we were assuming that the user would WANT to skip the manual approval entirely - including the notification. for example, if this was running against a pull request, would you want to send a slack notification every time the PR is updated? maybe but probably not (which is what i understand to be the premise of the use case in TEP-59)

maybe we can tweak the example a bit to make it work here - in this case, where we're running the pipeline on a PR, maybe we need to start a temporary cluster to deploy to? (since we don't want to deploy to prod)

  ```
          lint                     unit-tests
           |                           |
           v                           v
   report-linter-output        integration-tests
                                       |
                                       v
                                 manual-approval
                                 |            |
                                 v        (approver)
 (if pull-request)          build-image       |
   set-up-cluster                |             v
          |                      |        slack-msg
          v                      |
   (cluster ip, etc.).           |
          |                      v
          -------------->   deploy-image
  ```

In this example, when we run the pipeline for a pull request, we want to setup a cluster and pass the info about that cluster to deploy-image, but when we want to deploy to prod, we want to use params that are passed in with the prod cluster

(this is a bit flawed b/c you probably need more than just the ip and a string result might not be the best way to pass that - but maybe we could imagine in a world with dict params + results this makes a bit more sense?)

so we'd need to be able to express that when set-up-cluster is skipped, instead of trying to use it's results and skipping deploy-image (which depends on set-up-cluster), we use the params

Reply:
(or we could remove it, the next 2 use cases have it covered!)

Reply:
this is a useful use case, where the default results are variables that users would need to pass in as parameters

Reply:
Right, you have precisely captured a tweak [here](https://github.com/tektoncd/pipeline/pull/4085#discussion_r675893315) with the use case from TEP-0059.